Developer/Default.xctoolchain/cctools/man/check_dylib.1
.TH CHECK_DYLIB l "September 10, 2001" "Apple Computer, Inc."
.SH NAME
check_dylib \- perform checks on a dynamic library file
.SH SYNOPSIS
.B check_dylib
filename \-install_name
.I install_name
\-seg_addr_table
.I seg_addr_table
\-seg_addr_table_filename
.I path_name
.sp .5
.SH DESCRIPTION
.I Check_dylib
is a tool used by Apple's Build and Integration team to perform checks on a
Mach-O dynamic library as part of the build process.  The checks are performed
and
.IR check_dylib (l)
returns the status of the checks as exit values.
.PP
The checks are performed in the following order:
.PP
If the install name of the dynamic library does not start with @executable_path
check the install name of the dynamic library file against the specified
.B \-install_name
argument and if it does not match it exit with a value of 2.
.PP
Check the segment address table specified by
.B \-seg_addr_table
for an entry with the
.I path_name
argument of the
.B \-seg_addr_table_filename
and if one is not found in the table exit with a value of 3.
.PP
Check the dynamic library file's address against the address in the matching
entry in the segment address table specified by
.B \-seg_addr_table
and
.B \-seg_addr_table_filename
and if it does not match exit with a value of 4.
.PP
Check the address of the dynamic library and if it is zero exit with a value
of 5.
.PP
If there is any other errors
.IR check_dylib (l)
exits with a value of 1. If no checks fail then
.IR check_dylib (l)
exits with a value of 0.
.SH "SEE ALSO"
seg_addr_table(l), doug(0)


libs-gui/NEWS
1 NEWS
******

The currently released version of the library is '0.28.0'.

1.1 Noteworthy changes in version '0.28.0'
==========================================

This version adds support for modern XIB files and many new classes.
Plus the usual bunch of bug fixes.

   * Support loading of document XIB files.
   * Improve Key Value Binding for NSArrayController and add more
     bindings.
   * Better support for multi monitor usage and other improvement in the
     backend integration.
   * Add classes NSFontCollection, NSColorSampler, NSSpeechRecognizer,
     NSAppearance, NSPDFInfo, NSPICTImageRep, NSCIImageRep,
     NSPDFImageRep, NSPDFPanel, NSDataAsset, NSDatePicker,
     NSDatePickerCell, NSPredicateEditor, NSPredicateEditorRowTemplate,
     NSRuleEditor, NSGestureRecognizer, NSButtonTouchBarItem,
     NSCandidateListTouchBarItem, NSClickGestureRecognizer,
     NSColorPickerTouchBarItem, NSCustomTouchBarItem,
     NSGroupTouchBarItem, NSMagnificationGestureRecognizer,
     NSPanGestureRecognizer, NSPickerTouchBarItem,
     NSPopoverTouchBarItem, NSPressGestureRecognizer,
     NSRotationGestureRecognizer, NSSharingServicePickerTouchBarItem,
     NSSliderTouchBarItem, NSStepperTouchBarItem, NSTouchBarItem,
     NSTouchBar, NSTouch, NSDockTile.
   * Implement NSEPSImageRep.
   * Better encoding handling in RTF files.
   * Theming and drawing improvements.
   * Increase small font size to 10.
   * New cursor and stepper images.
   * Move NSFileWrapper to Foundation.
   * Fixed build on Debian GNU/kFreeBSD.
   * With command line argument -autolaunch YES, do not activate the
     application when -activateIgnoringOtherApps: is invoked.
   * Improvements to WindowMaker compatibility (e.g.  WMFHideApplication
     support).
   * Lowered NSFloatingWindowLevel by one to distinguish floating panels
     from menus.

1.2 Noteworthy changes in version '0.27.0'
==========================================

This version includes numerous bugfixes, compatibility improvements and
other changes accumulated over the last year.  It also enables work to
be done on integrating NSViews with a Core Animation renderer.

   * Make targetForAction safer.
   * Speed up menu updates.
   * Clean up speech tool compilation and switch to newer interface.
   * Fix bug in CUPS subclassing introduced in last release.
   * Minor improvements to typesetting.
   * Add NSIsControllerMarker.
   * Fix tracking on segmented cell.
   * Bring slider cell closer to Cocoa implementation.
   * Add ivar for Core Animation in NSView.
   * Improve border calculation on printing.
   * Lazy load app icon.
   * Better detection of removable volumes.
   * Polish translations.
   * Japanese translations.
   * Lots of bug fixes.

1.3 Noteworthy changes in version '0.26.2'
==========================================

This version is a small, but important bugfix release.

   * printing: Fix allocation of the CUPS printing classes.
   * installation: Fix the configure script.

1.4 Noteworthy changes in version '0.26.1'
==========================================

This version is released to conincide with version 1.25.1 of
gnustep-base, which contains changes required for this version of
gnustep-gui and gnustep-back.

   It includes an important workaround for users of GNUstep Objective-C
Runtime (libobjc2) and non-fragile ABI to avoid a bug in interaction
between the clang compiler and the runtime when non-fragile ABI is in
use.  Specifically, Clang and the runtime may disagree on what is the
offset of an ivar in a class's RAM. This manifested in a crash at
application startup due to misalignment of _gcontext inside NSThread.
See the mailing list discussion
(http://lists.gnu.org/archive/html/discuss-gnustep/2017-12/msg00129.html)
for more information.

   It also contains the following changes:

   * tests: Cleanup of warnings.
   * tests: Fix text system deallocation test.
   * printing: Undefine __BLOCKS__ before including cups.h, as some
     versions of the header expect that libdispatch is present and used
     if __BLOCKS__ is defined.
   * graphics context: Workaround for Clang+libobjc2+nonfragile ABI
     issue.

1.5 Noteworthy changes in version '0.26.0'
==========================================

This version was bumped due to previous binary incompatibilities between
0.25.0 and 0.25.1.  This version also includes numerous compatibility
improvements from the Summer of Code project, and a wide variety of
other fixes.  Notably, it fixes the use of cupsGetPPD() in the printing
system.

   * printing: Add an include to get deprecated function cupsGetPPD() on
     newer CUPS systems.
   * chore: Bump required base version.
   * tiff: Support for writing resolution.
   * jpeg: Save resolution information if it is different from 72 dpi.
   * save panel: Fix return type of sorting function.
   * events: Add some newer Cocoa enums and one method with dummy
     implementation.
   * speech synthesis: NSSpeechSynthesizerDelegate is now a @protocol on
     runtimes that support it.
   * pasteboard: New type identifiers.
   * translations: Some work on Polish, Russian and German translations
   * cell: Improvements to mouse tracking logic on NSCell.
   * image: If an unknown named image is unarchived with a coder or
     keyed coder, keep the name.
   * screen: Add -backingScaleFactor and return 1.0.
   * window: Return 1.0 from -backingScaleFactor.

   * compatibility: Numerous stub implementations of constants, classes
     and methods to improve source-level compatibility.
   * other bugfixes

1.6 Noteworthy changes in version '0.25.1'
==========================================

   * JPEG (saving) alpha channel fixes and size with resolution != 72
   * JPEG resolution read support
   * TIFF saving fixes
   * Improved volumes mounting and support
   * Portability improvements in volume mounting and support
   * Corrected layout of empty strings
   * Only update visible menus

1.7 Noteworthy changes in version '0.25.0'

Developer/Default.xctoolchain/cctools/misc/check_dylib.c
/*
 * Copyright (c) 2003 Apple Computer, Inc. All rights reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 * 
 * This file contains Original Code and/or Modifications of Original Code
 * as defined in and that are subject to the Apple Public Source License
 * Version 2.0 (the 'License'). You may not use this file except in
 * compliance with the License. Please obtain a copy of the License at
 * http://www.opensource.apple.com/apsl/ and read it before using this
 * file.
 * 
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
 * Please see the License for the specific language governing rights and
 * limitations under the License.
 * 
 * @APPLE_LICENSE_HEADER_END@
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include "stuff/ofile.h"
#include "stuff/errors.h"
#include "stuff/seg_addr_table.h"

struct check_block {
    char *install_name;
    enum bool check_result;
    struct seg_addr_table *entry;
};

/* used by error routines as the name of the program */
char *progname = NULL;

static void usage(
    void);

static void check_for_install_name(
    struct ofile *ofile,
    char *arch_name,
    void *cookie);

static void check_for_addresses(
    struct ofile *ofile,
    char *arch_name,
    void *cookie);

/* apple_version is created by the libstuff/Makefile */
extern char apple_version[];
char *version = apple_version;

/*
 * The check_dylib program.  It takes a dynamic library file, an -install_name
 * argument, a -seg_addr_table argument and a -seg_addr_table_filename argument.
 * Then it preforms the following checks in the following order and if the
 * specific check fails it returns a specific error code:
 *
 * Check:
 *	If the install_name of the dynamic library does not start with
 *	@executable_path checks the install_name of the dynamic library file
 *	against the specified -install_name argument and if it does not match it
 * Returns: 2
 *
 * Check:
 *	Checks the specified -seg_addr_table for the -seg_addr_table_filename
 *	and if not found in the table it
 * Returns: 3
 *
 * Check:
 *	Checks that the specified address in the -seg_addr_table for the
 *	-seg_addr_table_filename matches the dynamic library file.  If not it
 * Returns: 4
 *
 * Check:
 *	Checks that the specified address in the -seg_addr_table for the
 *	-seg_addr_table_filename and if it is zero then it
 * Returns: 5
 *
 * If there is any other errors it returns 1 (EXIT_FAILURE).  If all checks
 * pass then it returns 0 (EXIT_SUCCESS).
 */
int
main(
int argc,
char **argv,
char **envp)
{
    int i;
    uint32_t table_size;
    char *install_name, *image_file_name, *seg_addr_table_name,
         *seg_addr_table_filename;
    struct check_block block;
    struct seg_addr_table *seg_addr_table, *entry;

	progname = argv[0];
	install_name = NULL;
	image_file_name = NULL;
	seg_addr_table = NULL;
	seg_addr_table_filename = NULL;

	for(i = 1; i < argc; i++){
	    if(argv[i][0] == '-'){
		if(strcmp(argv[i], "-install_name") == 0){
		    if(i + 1 == argc){
			error("missing argument(s) to %s option", argv[i]);
			usage();
		    }
		    if(install_name != NULL){
			error("more than one: %s option", argv[i]);
			usage();
		    }
		    install_name = argv[i+1];
		    i++;
		}
		else if(strcmp(argv[i], "-seg_addr_table") == 0){
		    if(i + 1 == argc){
			error("missing argument(s) to %s option", argv[i]);
			usage();
		    }
		    if(seg_addr_table != NULL){
			error("more than one: %s option", argv[i]);
			usage();
		    }
		    seg_addr_table_name = argv[i+1];
		    seg_addr_table = parse_seg_addr_table(argv[i+1],
					  argv[i], argv[i+1], &table_size);
		    i++;
		}
		else if(strcmp(argv[i], "-seg_addr_table_filename") == 0){
		    if(i + 1 == argc){
			error("missing argument(s) to %s option", argv[i]);
			usage();
		    }
		    if(seg_addr_table_filename != NULL){
			error("more than one: %s option", argv[i]);
			usage();
		    }
		    seg_addr_table_filename = argv[i+1];
		    i++;
		}
		else{
		    error("unknown option %s\n", argv[i]);
		    usage();
		}
	    }
	    else{
		if(image_file_name != NULL){
		    error("more than file name specified (%s and %s)",
			  image_file_name, argv[i]);
		    usage();
		}
		image_file_name = argv[i];
	    }
	}
	if(image_file_name == NULL){
	    error("must specify a file name to be checked");
	    usage();
	}
	if(install_name == NULL){
	    error("must specify the -install_name <install_name> option");
	    usage();
	}
	if(seg_addr_table == NULL){
	    error("must specify the -seg_addr_table <table_name> option");
	    usage();
	}
	if(seg_addr_table_filename == NULL){
	    error("must specify the -seg_addr_table_filename <pathname> "			  "option");
	    usage();
	}

	/*
	 * The first check to perform is checking the install name to match
	 * the -install_name option.
	 */
	block.install_name = install_name;
	block.check_result = TRUE;
	ofile_process(image_file_name, NULL, 0, TRUE,
		      TRUE, TRUE, FALSE, check_for_install_name, &block);
	if(block.check_result == FALSE)
	    return(2);


	/*
	 * The next check to perform is to see if the -seg_addr_table_filename
	 *  has an entry in the specified -seg_addr_table.
	 */
	entry = search_seg_addr_table(seg_addr_table, seg_addr_table_filename);
	if(entry == NULL)
	    return(3);

	/*
	 * The next check to perform is to see if the address in the
	 * -seg_addr_table entry matches the dynamic library file.
	 */
	block.entry = entry;
	block.check_result = TRUE;
	ofile_process(image_file_name, NULL, 0, TRUE,
		      TRUE, TRUE, FALSE, check_for_addresses, &block);
	if(block.check_result == FALSE)
	    return(4);

	/*
	 * The next check to perform is to see address in the -seg_addr_table
	 * for the -seg_addr_table_filename is zero.
	 */
	if((entry->split == FALSE && entry->seg1addr == 0) ||
	   (entry->split == TRUE && (entry->segs_read_only_addr == 0 ||
				     entry->segs_read_write_addr == 0)) )
	    return(5);

	return(EXIT_SUCCESS);
}

/*
 * usage() prints the current usage message and exits indicating failure.
 */
static
void
usage(
void)
{
	fprintf(stderr, "Usage: %s <file_name> -install_name <install_name> "
		"-seg_addr_table <table_name> -seg_addr_table_filename "
		"<path_name>\n", progname);
	exit(EXIT_FAILURE);
}

static
void
check_for_install_name(
struct ofile *ofile,
char *arch_name,
void *cookie)
{
    uint32_t i;
    struct check_block *block;
    struct load_command *lc;
    struct dylib_command *dl;
    char *name;

#ifdef DEBUG
	printf("In check_for_install_name() ofile->file_name = %s",
	       ofile->file_name);
	if(arch_name != NULL)
	    printf(" arch_name = %s\n", arch_name);
	else
	    printf("\n");
#endif /* DEBUG */

	block = (struct check_block *)cookie;
	if(ofile->mh == NULL){
	    block->check_result = FALSE;
	    return;
	}

	lc = ofile->load_commands;
	for(i = 0; i < ofile->mh->ncmds; i++){
	    if(lc->cmd == LC_ID_DYLIB){
		dl = (struct dylib_command *)lc;
		name = (char *)lc + dl->dylib.name.offset;
		if(strncmp(name, "@executable_path",
		   sizeof("@executable_path") - 1) == 0)
		    return;
		if(strcmp(name, block->install_name) != 0)
		    block->check_result = FALSE;
		return;
	    }
	    lc = (struct load_command *)((char *)lc + lc->cmdsize);
	}
	block->check_result = FALSE;
	return;
}

static
void
check_for_addresses(
struct ofile *ofile,
char *arch_name,
void *cookie)
{
    uint32_t i, segs_read_only_addr, segs_read_write_addr;
    struct load_command *lc;
    struct segment_command *sg, *first;
    enum bool split;
    struct check_block *block;

#ifdef DEBUG
	printf("In check_for_addresses() ofile->file_name = %s",
	       ofile->file_name);
	if(arch_name != NULL)
	    printf(" arch_name = %s\n", arch_name);
	else
	    printf("\n");
#endif /* DEBUG */

	block = (struct check_block *)cookie;
	if(ofile->mh == NULL){
	    block->check_result = FALSE;
	    return;
	}

	split = (ofile->mh->flags & MH_SPLIT_SEGS) == MH_SPLIT_SEGS;
	if(block->entry->split != split){
	    block->check_result = FALSE;
	    return;
	}
	lc = ofile->load_commands;
	first = NULL;
	segs_read_only_addr = UINT_MAX;
	segs_read_write_addr = UINT_MAX;
	for(i = 0; i < ofile->mh->ncmds; i++){
	    if(lc->cmd == LC_SEGMENT){
		sg = (struct segment_command *)lc;
		if(first == NULL){
		    first = sg;
		    if(split == FALSE &&
		       first->vmaddr != block->entry->seg1addr){
			block->check_result = FALSE;
			return;
		    }
		}
		if((sg->initprot & VM_PROT_WRITE) == 0){
		    if(split == TRUE && sg->vmaddr < segs_read_only_addr)
			segs_read_only_addr = sg->vmaddr;
		}
		else{
		    if(split == TRUE && sg->vmaddr < segs_read_write_addr)
			segs_read_write_addr = sg->vmaddr;
		}
	    }
	    lc = (struct load_command *)((char *)lc + lc->cmdsize);
	}
	if(split == TRUE){
	    if(segs_read_only_addr != block->entry->segs_read_only_addr ||
	       segs_read_write_addr != block->entry->segs_read_write_addr){
		block->check_result = FALSE;
		return;
	    }
	}
}
